docs/explanations/rendering.md

```diff
@@ -0,0 +1,344 @@
+other decisions with partially loaded data, so the user can work faster. Also
+progressive rendering just feels more pleasant for the user. Progressive rending
+makes the internal state of the application visible which is often leads to a
+better user experience.
+
+## Chunked File Formats
+
+The most common types of chunks are blocks of contiguous memory inside a chunked
+file format like **Zarr** and exposed by an API like **Dask**. If an
+image is stored without chunks then reading a 2D rectangle would require
+hundreds of small read operations from all over the file. With chunks you can
+read a rectangular region with a single read operation.
+
+![chunked-format](images/chunked-format.png)
+
+For 3D images the chunks tend to be 3D blocks, but the idea is the same.
+Neuroglancer commonly stores the data in 64x64x64 voxel chunks which is 0.25MB
+per chunk. This is useful because you can read the data in XY, XZ or YZ and it
+performs the same in each orientation. It's also nice because you scroll through
+slices quickly since on average you have 32 slices above and below your current
+location.
+
+## Creating Chunks
+
+In #1300 there are no chunks, the images were created in memory as one
+monolithic array. So to solve #1300 we are going to have to break that array
+into chunks in order to send it to the graphics card incrementally. In #1320 the
+images are small so we are not chunked, but there are 3 image layers, so we can
+consider the full layers to be chunks. In general we can get creative with
+chunks, they can be spatial subdivisions or any other division we want.
+
+With non-image data like points, shapes and meshes we can have 2D or 3D spatial
+chunks, we can have layers, and we can invent other sub-divisions to use as
+chunks. *As long as things can be loaded and drawn independently we can use them
+as chunks*.
+
+# Loading into RAM and VRAM
+
+There is a two step process to get data into VRAM where we can draw it. First it
+needs to be loaded into RAM and then transferred into VRAM.
+
+Loading into RAM must be done in a thread since we don't know how long it will
+take. For example loading data over the internet or doing a complex calculation
+to produce the data could both take really long time. We are going to use the
+new `@thread_worker` interface for our thread pool.
+
+Loading into VRAM is a different story because it must happen in the GUI thread,
+at least that is our assumption while we are using OpenGL. Therefore we need to
+amortize the load over some number of frames. We will set a budget, for
+example 5 milliseconds. Each frame can spend that much time loading data into
+VRAM, it will spend the rest of the frame drawing as normal. In this case it's
+important that no single chunk takes more than 5 milliseconds to transfer.
+
+![paging-chunks](images/paging-chunks.png)
+
+Viewed as a timeline the rendering thread render regularly spaced frames while
```

Comment:
```suggestion
Viewed as a timeline, the rendering thread renders regularly spaced frames, while
```
?

Reply:
Fixed to say "Viewed as a timeline, the rendering thread has regularly spaced frames, while the IO and compute threads load data into RAM in parallel.